File: scripts/algorithms/C/Construct Binary Tree from Preorder and Inorder Traversal/Construct Binary Tree from Preorder and Inorder Traversal.py
# Runtime: 274 ms (Top 36.07%) | Memory: 88.6 MB (Top 22.22%)

# Definition for a binary tree node.
# class TreeNode(object):
# def __init__(self, val=0, left=None, right=None):
# self.val = val
# self.left = left
# self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if not preorder or not inorder:
            return None

        root = TreeNode(preorder[0])
        mid = inorder.index(preorder[0])
        root.left = self.buildTree(preorder[1:mid+1], inorder[:mid])
        root.right = self.buildTree(preorder[mid+1:], inorder[mid+1:])
        return root

File: scripts/algorithms/C/Construct Binary Tree from Preorder and Inorder Traversal/Construct Binary Tree from Preorder and Inorder Traversal.js
// Runtime: 121 ms (Top 79.59%) | Memory: 44.5 MB (Top 97.46%)
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 * this.val = (val===undefined ? 0 : val)
 * this.left = (left===undefined ? null : left)
 * this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    let idx = 0;
    function create(inorderStart, inorderEnd) {
        if (idx === preorder.length || inorderStart > inorderEnd) {
            return null;
        }
        const root = new TreeNode(preorder[idx]);
        const mid = inorder.indexOf(preorder[idx]);
        idx++;
        root.left = create(inorderStart, mid -1);
        root.right = create(mid + 1, inorderEnd);
        return root;
    }

    return create(0, inorder.length -1);
};